src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,6 +307,115 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
+            // If Right-to-left is ever added to the source generator, then we would need to the logic to define
+            // bump, as well as stoppos
+            Debug.Assert((options & RegexOptions.RightToLeft) == 0);
+
+            EmitPrevLenCheck();
+            writer.WriteLine();
+
+            EmitMainSearchLoop();
+            writer.WriteLine();
+
+            return;
+
+#pragma warning disable CS8321 // Local function is declared but never used
+            void EmitTimeoutHandling()
+            {
+                writer.WriteLine("// Handle timeout argument");
+                writer.WriteLine("_timeout = -1;");
+                writer.WriteLine("bool ignoreTimeout = global::System.Text.RegularExpressions.Regex.InfiniteMatchTimeout == timeout;");
+                using (EmitBlock(writer, "if (!ignoreTimeout)"))
+                {
+                    writer.WriteLine("// We are using Environment.TickCount and not Stopwatch for performance reasons.");
+                    writer.WriteLine("// Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt");
+                    writer.WriteLine("// overflow it will still stay ahead of Environment.TickCount for comparisons made");
+                    writer.WriteLine("// in DoCheckTimeout().");
+                    writer.WriteLine("global::System.Text.RegularExpressions.Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected");
+                    writer.WriteLine("_timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;");
+                    writer.WriteLine("_timeoutOccursAt = global::System.Environment.TickCount + _timeout;");
+                    writer.WriteLine("_timeoutChecksToSkip = TimeoutCheckFrequency;");
+                }
+            }
+#pragma warning restore CS8321 // Local function is declared but never used
+
+            void EmitPrevLenCheck()
+            {
+                writer.WriteLine("// If previous match was empty or failed, advance by one before matching.");
+                using (EmitBlock(writer, "if (prevlen == 0)"))
+                {
+                    using (EmitBlock(writer, "if (textstart == text.Length)"))
+                    {
+                        writer.WriteLine("base.runmatch = global::System.Text.RegularExpressions.Match.Empty;");
+                        writer.WriteLine("return;");
+                    }
+                    writer.WriteLine();
+                    writer.WriteLine("base.runtextpos++;");
+                }
+            }
+
+            void EmitMainSearchLoop()
+            {
+                using (EmitBlock(writer, "while (true)"))
+                {
+                    using (EmitBlock(writer, "if (FindFirstChar(text))"))
+                    {
+                        writer.WriteLine("base.CheckTimeout();");
```

Comment:
We'll want to conditionalize this as we do today with other timeout checks in Go.